### PR TITLE
Provide a default resolver and allow resolvers to be configured to disallow resolution from the operating system

### PR DESCRIPTION
**Pull request by mattrm456**
This PR adds a variety of functions to `ntsf::System` to support the installation, retrieval, and usage of a "default" blocking resolver. The goals of such "default" objects are:

1) They are globally accessible and enable usage of the functionality provided by the object without having to (or requiring the ability to) pass an object as a parameter through chains of functions.

2) They allow `main()` to configure a default behavior of these objects, given the context of the program, useful (but not limited to) test drivers.

This PR is the beginning of a planned series of future PRs to add support for "default" objects of a variety of other abstract classes in NTS and NTC. For example, `ntci::Reactor`, `ntci::Interface`, etc.

The following example illustrates how to install a default resolver configured to only return a set of pre-determined resolutions. First, install a resolver as the default resolver.
```
ntsa::ResolverConfig resolverConfig;
resolverConfig.setOverridesEnabled(true);
resolverConfig.setSystemEnabled(false);

bsl::shared_ptr<ntsi::Resolver> resolver = ntsf::System::createResolver(
    resolverConfig,
    bslma::Default::globalAllocator());

ntsf::System::setDefault(resolver);
```
Next, assert the default resolver matches the resolver we installed.
```
bsl::shared_ptr<ntsi::Resolver> resolverDefault;
ntsf::System::getDefault(&resolverDefault);

BSLS_ASSERT(resolverDefault == resolver);
```
Finally, use the static functions to specify an override for the IP address assigned to a domain name, resolve that domain name, and assert the result matches the IP address we specified.
```
ntsf::System::addIpAddress("example.com", ntsa::IpAddress("1.2.3.4"));

bsl::vector<ntsa::IpAddress> ipAddressList;
ntsf::System::getIpAddress(&ipAddressList, "example.com", ntsa::IpAddressOptions());
        
BSLS_ASSERT(ipAddressList[0] == ntsa::IpAddress("1.2.3.4"));
```

### CODE REVIEW

File: groups/nts/ntsf/ntsf_system.h [hunk 1]
```diff
@@ -895,6 +896,13 @@ struct System {
     static bsl::shared_ptr<ntsi::Resolver> createResolver(
         bslma::Allocator* basicAllocator = 0);
 
+    /// Create a new resolver having the specified 'configurationn'. Optionally
```
Comment (smtrfnv) [inline groups/nts/ntsf/ntsf_system.h:899]: `configuration`

**Review by smtrfnv — APPROVED** [review]

File: groups/nts/ntsa/ntsa_resolverconfig.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+class ResolverConfig
+{
+    bdlb::NullableValue<bool> d_overridesEnabled;
+    bdlb::NullableValue<bool> d_systemEnabled;
+
+  public:
+    /// Create a new resolver configuration.
+    ResolverConfig();
+
+    /// Create  anew resolver configuration having the same value as the
```
[the remark this answers is not in the document]
Comment (mattrm456) [reply groups/nts/ntsa/ntsa_resolverconfig.h:58]: Done.

File: groups/nts/ntsf/ntsf_system.h (hunk 1 above)
Comment (mattrm456) [reply groups/nts/ntsf/ntsf_system.h:899]: Done.

File: groups/nts/ntsf/ntsf_system.cpp [hunk 3]
```diff
@@ -44,7 +56,19 @@ ntsa::Error System::initialize()
         if (rc != 0) {
             return ntsa::Error(ntsa::Error::last());
         }
+
+        // We use a new delete allocator instead of the global allocator here
+        // because we want prevent a visible "memory leak" if the global
```
[the remark this answers is not in the document]
Comment (mattrm456) [reply groups/nts/ntsf/ntsf_system.cpp:61]: Done.

**Pull request merged**